temporalcli/commands_test.go

```diff
@@ -47,13 +50,17 @@ func (h *CommandHarness) Close() {
 	}
 }
 
+// Pieces must appear in order on the line and not overlap
 func (h *CommandHarness) ContainsOnSameLine(text string, pieces ...string) {
 	// Split into lines, then check each piece is present
 	lines := strings.Split(text, "\n")
 	for _, line := range lines {
 		foundAll := true
+		lastEndIndex := -1
 		for _, piece := range pieces {
-			if !strings.Contains(line, piece) {
+			if index := strings.Index(line, piece); index > lastEndIndex {
+				lastEndIndex = index + len(piece)
+			} else {
```

Comment:
Ha, algorithms like this are hard to get right first time without unit tests! I suggest we use a regular-expression based implementation, i.e. form a regex by joining `pieces` with whatever separator that gives the separator semantics that we think callers will want in the common case. That way the behavior of the utility is clear to anyone who looks at the code, and they can use their own regex assertion if they want different behavior. I'd also be OK with deleting this utility and requiring callers to use a regex explicitly.

As it is it still has bugs. The most serious I'm seeing is that `ContainsOnSameLine("a b a", "b", "a")` will fail, when it should pass.

But also `ContainsOnSameLine("ab", "a", "b")` will fail (no gap between them) and yet `ContainsOnSameLine("axb", "a", "b")` will pass.

And related to the first bug, `ContainsOnSameLine("a a", "a", "a")` will fail.

Reply:
This is why I wanted it simple and ask users to use regex if they needed more instead of increase the rules of this little helper. I regret not sticking to the original simple/looser need. At this point I think we should just document the limits and tell people to use regex if they want.

Reply:
The original version allowed the pieces to be in any order and overlapping, which isn't appropriate for a test harness assertion utility.

What do you see as the downside of implementing `ContainsOnSameLine` as a regex search for `pieces` joined by something like `.*` or `.+` or `.*\s.*`? However it's implemented, the resulting function will have _some_ behavior, which we need to make clear to callers. A regex is the best way to make that clear to callers as I see it.

Or, as I mentioned above, we could just delete the utility and let callers decide on their regex.

Reply:
> What do you see as the downside of implementing ContainsOnSameLine as a regex search for pieces joined by something like .* or .+ or .*\s.*?

The continual requirements. "Can you strip colors too?", "Can you make it case insensitive?", "Can you make it only allow whole words instead of partial match?", "Can you allow this to work on adjacent lines instead of just one?", etc etc. "But it isn't appropriate for an stdout test harness utility to not strip out ANSI codes", "But it isn't appropriate for this test harness utility to not check whole words", etc. Each has their own opinion and it was just meant to be a little code reuse thing, but I didn't want to name it `ContainsAnywhereOnSameLineInAnyOrderMayOverlap`.

> Or, as I mentioned above, we could just delete the utility and let callers decide on their regex.

This may be best, but I still want some code reuse and I accept the limitations of the little helper.

Reply:
> The continual requirements. "Can you strip colors too?", "Can you make it case insensitive?", "Can you make it only allow whole words instead of partial match?", "Can you allow this to work on adjacent lines instead of just one?", etc etc. "But it isn't appropriate for an stdout test harness utility to not strip out ANSI codes", "But it isn't appropriate for this test harness utility to not check whole words", etc.

My suggestions that a string matching function should require order and not allow overlapping matches are perfectly reasonable and uncontroversial and it's not appropriate to attempt to ridicule them like that.

Reply:
Updated to support regex-based match

Reply:
Thanks, that looks great to me.